Code/src/main/java/nl/utwente/group10/haskell/expr/Apply.java

```diff
@@ -1,66 +1,60 @@
 package nl.utwente.group10.haskell.expr;
 
+import nl.utwente.group10.haskell.env.Env;
 import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
 import nl.utwente.group10.haskell.type.FuncT;
 import nl.utwente.group10.haskell.type.Type;
 
-import java.util.Arrays;
-
 /**
- * A type of expression that is a lazy application of a function with certain arguments.
+ * Lazy application of an argument to a function.
```

Comment:
Nothing is evaluated. Until you call analyze(), we don't even know if the application is valid. I would call that pretty lazy. So either ```Apply``` is lazy, or it should get its ass kicked.